idaes/gas_solid_contactors/properties/oxygen_iron_OC_oxidation/gas_phase_thermo.py

```diff
@@ -0,0 +1,798 @@
+                calculate_variable_from_constraint(
+                            blk[k].mw,
+                            blk[k].mw_eqn)
+
+            if hasattr(blk[k], "ideal_gas"):
+                calculate_variable_from_constraint(
+                            blk[k].dens_mol,
+                            blk[k].ideal_gas)
+
+            if hasattr(blk[k], "dens_mass_basis"):
+                calculate_variable_from_constraint(
+                            blk[k].dens_mass,
+                            blk[k].dens_mass_basis)
+
+            if hasattr(blk[k], "mixture_heat_capacity_eqn"):
+                calculate_variable_from_constraint(
+                            blk[k].cp_mol,
+                            blk[k].mixture_heat_capacity_eqn)
+
+            if hasattr(blk[k], "cp_mass_basis"):
+                calculate_variable_from_constraint(
+                            blk[k].cp_mass,
+                            blk[k].cp_mass_basis)
+
+            if hasattr(blk[k], "visc_d_constraint"):
+                calculate_variable_from_constraint(
+                            blk[k].visc_d,
+                            blk[k].visc_d_constraint)
+
+            if hasattr(blk[k], "therm_cond_constraint"):
+                calculate_variable_from_constraint(
+                            blk[k].therm_cond,
+                            blk[k].therm_cond_constraint)
+
+            if hasattr(blk[k], "mixture_enthalpy_eqn"):
+                calculate_variable_from_constraint(
+                            blk[k].enth_mol,
+                            blk[k].mixture_enthalpy_eqn)
+
+            for j in blk[k]._params.component_list:
+
+                if hasattr(blk[k], "comp_conc_eqn"):
+                    calculate_variable_from_constraint(
+                                blk[k].dens_mol_comp[j],
+                                blk[k].comp_conc_eqn[j])
+
+                if hasattr(blk[k], "diffusion_comp_constraint"):
+                    calculate_variable_from_constraint(
+                                blk[k].diffusion_comp[j],
+                                blk[k].diffusion_comp_constraint[j])
+
+                if hasattr(blk[k], "cp_shomate_eqn"):
+                    calculate_variable_from_constraint(
+                        blk[k].cp_mol_comp[j],
+                        blk[k].cp_shomate_eqn[j])
+
+                if hasattr(blk[k], "enthalpy_shomate_eqn"):
+                    calculate_variable_from_constraint(
+                            blk[k].enth_mol_comp[j],
+                            blk[k].enthalpy_shomate_eqn[j])
+
+        # Solve property block if non-empty
+        free_vars = 0
+        for k in blk.keys():
+            free_vars += number_unfixed_variables_in_activated_equalities(
+                blk[k])
+
+        if free_vars > 0:
+            with idaeslog.solver_log(solve_log, idaeslog.DEBUG) as slc:
+                res = solve_indexed_blocks(opt, [blk], tee=slc.tee)
+        else:
```

Comment:
Good question. The variables not initialized by `calculate_variables_from_constraint` will take on their `initialize` values which might be arbitrary and not feasible for the problem to be solved. However, a quick check shows that not solving the blocks might shave some time of the solve if the current (prior to solving the block) values of the variables are feasible. I don't use `calculate_variables_from_constraint` for all the variables as, depending on the (nonlinearity of the) constraint, it can be finicky and give infeasible local solutions.

Reply:
If `calculate_variable_from_constraint` is failing, I think that means we are calculating variables in the wrong order. It should never fail if the variable we are solving for are linear in its constraints and the other variables are initialized. Determining the right order at runtime, however, requires a block triangularization, which could be implemented but requires a bit of thought about how to do it right (e.g. how to avoid repeating work, what to do in edge cases,...).